finite-state/shared/src/main/scala/fs2/data/mft/MFT.scala

```diff
@@ -65,7 +66,7 @@ private[data] class MFT[InTag, OutTag](init: Int, rules: Map[Int, Rules[InTag, O
     * The generated ESP contains one decision tree encoding all the patterns
     * of this MFT.
     */
-  def esp[F[_]](implicit F: Sync[F]): F[ESP[F, InTag, OutTag]] = {
+  def esp[F[_]](implicit F: MonadError[F, Throwable] with Defer[F]): F[ESP[F, InTag, OutTag]] = {
```

Comment:
```suggestion
  def esp[F[_]](implicit F: MonadError[F, Throwable], defer: Defer[F]): F[ESP[F, InTag, OutTag]] = {
```

I would not rely on both being implemented in the same place.